### PR TITLE
Feat: Added codeSamples and propsMethods settings per sections

### PR DESCRIPTION
**Pull request by rafaesc**
First PR to abstract the sections setting, related with https://github.com/styleguidist/react-styleguidist/issues/892

### CODE REVIEW

File: src/utils/getPageTitle.js [hunk 1]
```diff
@@ -12,9 +12,12 @@ import { DisplayModes } from '../consts';
...
 export default function getPageTitle(sections, baseTitle, displayMode) {
-	if (displayMode === DisplayModes.component || displayMode === DisplayModes.example) {
+	if (
+		displayMode === DisplayModes.component ||
+		(displayMode === DisplayModes.example && sections[0].components.length)
+	) {
 		return `${sections[0].components[0].name} — ${baseTitle}`;
-	} else if (displayMode === DisplayModes.section) {
+	} else if (displayMode === DisplayModes.section || displayMode === DisplayModes.example) {
 		return `${sections[0].name} — ${baseTitle}`;
```
Comment (rafaesc) [inline src/utils/getPageTitle.js]: Fix #943

File: docs/Configuration.md [hunk 2]
```diff
@@ -664,7 +664,7 @@ module.exports = {
 
 > **Note:** `entry`, `externals`, `output`, `watch`, and `stats` options will be ignored. For production builds, `devtool` will also be ignored.
 
-> **Note:** `CommonsChunkPlugins`, `HtmlWebpackPlugin`, `UglifyJsPlugin`, `HotModuleReplacementPlugin` plugins will be ignored because Styleguidist already includes them or they may break Styleguidist.
+> **Note:** `CommonsChunkPlugins`, `HtmlWebpackPlugin`, `MiniHtmlWebpackPlugin`, `UglifyJsPlugin`, `HotModuleReplacementPlugin` plugins will be ignored because Styleguidist already includes them or they may break Styleguidist.
```
Comment (sapegin) [inline docs/Configuration.md]: Looks like some merge issue — these changes are from another PR ;-/

File: src/rsg-components/Playground/Playground.js [hunk 3]
```diff
@@ -63,10 +65,11 @@ export default class Playground extends Component {
 
 	render() {
 		const { code, activeTab } = this.state;
-		const { evalInContext, index, name, settings } = this.props;
+		const { evalInContext, index, name, settings, codeSamples } = this.props;
 		const { displayMode } = this.context;
+		const hideEditor = codeSamples === CodeSamplesModes.hide;
 		const preview = <Preview code={code} evalInContext={evalInContext} />;
-		if (settings.noeditor) {
+		if (settings.noeditor || hideEditor) {
```
Comment (okonet) [inline src/rsg-components/Playground/Playground.js]: Can we move both conditions to the const?

File: src/rsg-components/Playground/Playground.js [hunk 4]
```diff
@@ -63,10 +65,11 @@ export default class Playground extends Component {
 
 	render() {
 		const { code, activeTab } = this.state;
-		const { evalInContext, index, name, settings } = this.props;
+		const { evalInContext, index, name, settings, codeSamples } = this.props;
 		const { displayMode } = this.context;
+		const hideEditor = codeSamples === CodeSamplesModes.hide;
```
Comment (okonet) [inline src/rsg-components/Playground/Playground.js]: If `hideEditor` is `true` we should not render `preview` since it will take time. Let put it into the condition below.

File: src/utils/getRouteData.js [hunk 5]
```diff
@@ -46,9 +46,9 @@ export default function getRouteData(sections, hash, pagePerSection) {
 
 	// Filter the requested component if required
 	if (targetName) {
-		const filteredComponents = filterComponentsInSectionsByExactName(sections, targetName);
-		if (filteredComponents.length) {
-			sections = [{ components: filteredComponents }];
+		const filteredSections = filterComponentsInSectionsByExactName(sections, targetName);
```
Comment (okonet) [inline src/utils/getRouteData.js]: What are these changes about?

Comment (rafaesc) [reply src/utils/getRouteData.js]: yes, it's because in this case tries isolated so if I want that use the same section setting, its necessary that the sections have `codeSamples` and `propsMethods` if you see on the function `filterComponentsInSectionsByExactName`, right now returns a section.
https://github.com/styleguidist/react-styleguidist/pull/941/files#diff-5cafb86def7c61e87a06ed3f7c5b95cc

File: src/rsg-components/Playground/Playground.js (hunk 4 above)
Comment (rafaesc) [reply src/rsg-components/Playground/Playground.js]: But I have understood that `preview` always should render, like this
https://github.com/styleguidist/react-styleguidist/blob/master/src/rsg-components/Playground/Playground.js#L68-L72, or how would I have to create the condition?, Thanks!

**Comment by viljamis** [comment]
@rafaesc Nice work! Let me know if I can help in any way here.

File: src/rsg-components/Components/Components.spec.js [hunk 6]
```diff
@@ -3,6 +3,8 @@ import ReactComponent from '../ReactComponent';
 import Components from './Components';
 import ComponentsRenderer from './ComponentsRenderer';
 
+const codeSamples = 'collapse';
```
Comment (sapegin) [inline src/rsg-components/Components/Components.spec.js]: Shouldn't it be `CodeSamplesModes.collapse`?

File: src/consts.js [hunk 7]
```diff
@@ -11,3 +11,15 @@ export const DisplayModes = Object.freeze({
 	example: 'example',
 	// TODO: error (404)
 });
+
+export const CodeSamplesModes = Object.freeze({
```
Comment (sapegin) [inline src/consts.js]: I think `ExampleModes` would be better.

File: src/consts.js [hunk 8]
```diff
@@ -11,3 +11,15 @@ export const DisplayModes = Object.freeze({
...
+	hide: 'hide',
+	collapse: 'collapse',
+	expand: 'expand',
+});
+
+export const PropsMethodsModes = Object.freeze({
```
Comment (sapegin) [inline src/consts.js]: And `UsageModes`, we have the `Usage` component to render props and methods.

File: src/utils/__tests__/filterComponentsInSectionsByExactName.spec.js [hunk 9]
```diff
@@ -23,6 +23,6 @@ const sections = deepfreeze([
 describe('filterComponentsInSectionsByExactName', () => {
 	it('should return components at any level with exact name', () => {
 		const result = filterComponentsInSectionsByExactName(sections, 'Image');
-		expect(result.map(x => x.name)).toEqual(['Image']);
+		expect(result[0].components.map(x => x.name)).toEqual(['Image']);
```
Comment (sapegin) [inline src/utils/__tests__/filterComponentsInSectionsByExactName.spec.js]: What was the reason for this change? Suspicious ;-)

File: src/utils/filterComponentsInSectionsByExactName.js [hunk 10]
```diff
@@ -5,17 +5,24 @@ import filterComponentsByExactName from './filterComponentsByExactName';
  *
  * @param {object} sections
  * @param {string} name
- * @return {Array}
+ * @return {Array} sections
```
Comment (sapegin) [inline src/utils/filterComponentsInSectionsByExactName.js]: Return shouldn't have a variable name.

File: src/utils/filterComponentsInSectionsByExactName.js [hunk 11]
```diff
@@ -5,17 +5,24 @@ import filterComponentsByExactName from './filterComponentsByExactName';
...
  */
 export default function filterComponentsInSectionsByExactName(sections, name) {
-	const components = [];
+	const output = [];
```
Comment (sapegin) [inline src/utils/filterComponentsInSectionsByExactName.js]: `output` is a very non-descriptive name. If you want to use `components` inside the loop, use something like `filteredComponents` instead of `output`.

File: src/utils/__tests__/filterComponentsInSectionsByExactName.spec.js (hunk 9 above)
Comment (rafaesc) [reply src/utils/__tests__/filterComponentsInSectionsByExactName.spec.js]: Yes, because before returned a component list to isolate them, but how I need the `codeSample` and `propsMethods` section setting to mantain the setting, now it returns section list

**Comment by sapegin** [comment]
Do we have any breaking changes here? Do we need any docs updates (I guess we do)?

File: docs/Components.md [hunk 12]
```diff
@@ -109,6 +109,8 @@ Each section consists of (all fields are optional):
 * `components` — a glob pattern string, an array of component paths or a function returning a list of components. The same rules apply as for the root `components` option.
 * `sections` — array of subsections (can be nested).
 * `description` — A small description of this section.
+* `codeSamples` — Settings documentation initially of example code tab, uses [codeSamples](Configuration.md#codesamples).
```
Comment (sapegin) [inline docs/Components.md]: “Initial state of the code example tab”? And similarly the next one.

File: docs/Configuration.md [hunk 13]
```diff
@@ -55,6 +55,16 @@ Type: `String`, optional
 
 Your application static assets folder, will be accessible as `/` in the style guide dev server.
 
+#### `codeSamples`
```
Comment (sapegin) [inline docs/Configuration.md]: They are _examples_, not samples — I think I’ve mentioned that before ;-)

File: docs/Configuration.md [hunk 14]
```diff
@@ -55,6 +55,16 @@ Type: `String`, optional
...
+
+Type: `String`, default: `collapse`
+
+Allows to config documentation initially of example code tab
```
Comment (sapegin) [inline docs/Configuration.md]: Defines the initial state of…

File: docs/Configuration.md [hunk 15]
```diff
@@ -55,6 +55,16 @@ Type: `String`, optional
...
+
+* when is `collapse`: collapses the tab by default.
```
Comment (sapegin) [inline docs/Configuration.md]: I'd remove “when is”.

File: docs/Configuration.md [hunk 16]
```diff
@@ -664,7 +672,7 @@ module.exports = {
 
 > **Note:** `entry`, `externals`, `output`, `watch`, and `stats` options will be ignored. For production builds, `devtool` will also be ignored.
 
-> **Note:** `CommonsChunkPlugins`, `HtmlWebpackPlugin`, `MiniHtmlWebpackPlugin`, `UglifyJsPlugin`, `HotModuleReplacementPlugin` plugins will be ignored because Styleguidist already includes them or they may break Styleguidist.
+> **Note:** `CommonsChunkPlugins`, `HtmlWebpackPlugin`, `UglifyJsPlugin`, `HotModuleReplacementPlugin` plugins will be ignored because Styleguidist already includes them or they may break Styleguidist.
```
Comment (sapegin) [inline docs/Configuration.md]: Ooops ;-/

File: docs/Configuration.md (hunk 14 above)
Comment (sapegin) [inline docs/Configuration.md]: Defines the initial state of the example code tab:

File: docs/Configuration.md (hunk 13 above)
Comment (sapegin) [inline docs/Configuration.md]: They are _examples_, not samples — I think I’ve mentioned that before ;-)

File: docs/Configuration.md [hunk 17]
```diff
@@ -335,6 +345,16 @@ module.exports = {
 }
 ```
 
+#### `propsMethods`
```
Comment (sapegin) [inline docs/Configuration.md]: And I’d keep “usage” here.

File: docs/Webpack.md [hunk 18]
```diff
@@ -43,7 +43,7 @@ module.exports = {
 
 > **Note:** `entry`, `externals`, `output`, `watch`, and `stats` options will be ignored. For production builds, `devtool` will also be ignored.
 
-> **Note:** `CommonsChunkPlugins`, `HtmlWebpackPlugin`, `MiniHtmlWebpackPlugin`, `UglifyJsPlugin`, `HotModuleReplacementPlugin` plugins will be ignored because Styleguidist already includes them or they may break Styleguidist.
+> **Note:** `CommonsChunkPlugins`, `HtmlWebpackPlugin`, `UglifyJsPlugin`, `HotModuleReplacementPlugin` plugins will be ignored because Styleguidist already includes them or they may break Styleguidist.
```
Comment (sapegin) [inline docs/Webpack.md]: Some unrelated changes in this file — probably merge issue.
